File: go/155.go
package main

import "math"

type MinStack struct {
	st  []int
	min int
}

func Constructor() MinStack {
	return MinStack{make([]int, 0), math.MaxInt32}
}

func (this *MinStack) Push(val int) {
	if val <= this.min {
		this.st = append(this.st, this.min)
		this.min = val
	}
	this.st = append(this.st, val)
}

func (this *MinStack) Pop() {
	if this.st[len(this.st)-1] == this.GetMin() {
		this.min = this.st[len(this.st)-2]
		this.st = this.st[:len(this.st)-2]
	} else {
		this.st = this.st[:len(this.st)-1]
	}
}

func (this *MinStack) Top() int {
	return this.st[len(this.st)-1]
}

func (this *MinStack) GetMin() int {
	return this.min
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: python/155.py
from typing import List


class MinStack:

    def __init__(self):
        self.st: List = []
        self.m: int = 2**31-1

    def push(self, val: int) -> None:
        self.st.append(self.m)
        self.st.append(val)
        self.m = min(self.m, val)

    def pop(self) -> None:
        self.st.pop(-1)
        self.m = self.st[-1]
        self.st.pop(-1)

    def top(self) -> int:
        return self.st[-1]

    def getMin(self) -> int:
        return self.m


# Your MinStack object will be instantiated and called as such:
# obj = inStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
